File: frontend/lib/llm-scorer.ts
/**
 * LLM Flight Scorer
 * 
 * Uses an LLM to evaluate and score FlightOffers against user preferences.
 * Returns scored offers with reasoning and match details.
 */

import { generateObject } from "ai"
import { openai } from "@ai-sdk/openai"
import { z } from "zod"
import type { FlightOffer, FlightPreferences, ScoredFlightOffer } from "./types"

const ScoringSchema = z.object({
  score: z.number().min(0).max(100).describe("Overall score from 0-100"),
  reasoning: z.string().describe("Explanation of why this score was given"),
  match_details: z.object({
    budget_match: z.number().min(0).max(1).optional().describe("How well it matches budget (0-1)"),
    time_match: z.number().min(0).max(1).optional().describe("How well departure/arrival times match (0-1)"),
    layover_match: z.number().min(0).max(1).optional().describe("How well layovers match preferences (0-1)"),
    airline_match: z.number().min(0).max(1).optional().describe("How well airlines match preferences (0-1)"),
    overall_fit: z.number().min(0).max(1).optional().describe("Overall fit score (0-1)"),
  }),
})

type ScoringResult = z.infer<typeof ScoringSchema>

/**
 * Score a single flight offer against user preferences
 */
export async function scoreFlightOffer(
  offer: FlightOffer,
  preferences: FlightPreferences,
): Promise<ScoredFlightOffer> {
  const prompt = buildScoringPrompt(offer, preferences)

  try {
    const { object: scoring } = await generateObject({
      model: openai("gpt-4o-mini"), // Use OpenAI provider directly
      schema: ScoringSchema,
      prompt,
      temperature: 0.3,
    })

    return {
      ...offer,
      score: scoring.score,
      reasoning: scoring.reasoning,
      match_details: scoring.match_details,
    }
  } catch (error) {
    console.error("[LLM Scorer] Error scoring flight offer:", error)
    // Return fallback score based on simple rules
    return scoreFlightOfferFallback(offer, preferences)
  }
}

/**
 * Score multiple flight offers in batch
 * Uses parallel processing for efficiency
 */
export async function scoreFlightOffers(
  offers: FlightOffer[],
  preferences: FlightPreferences,
): Promise<ScoredFlightOffer[]> {
  // Score in parallel (be mindful of rate limits)
  const scoredOffers = await Promise.all(
    offers.map((offer) => scoreFlightOffer(offer, preferences).catch((error) => {
      console.error(`[LLM Scorer] Error scoring offer ${offer.id}:`, error)
      return scoreFlightOfferFallback(offer, preferences)
    }))
  )

  // Sort by score (highest first)
  return scoredOffers.sort((a, b) => b.score - a.score)
}

function buildScoringPrompt(offer: FlightOffer, preferences: FlightPreferences): string {
  let prompt = `You are a flight evaluation expert. Score this flight offer against the user's preferences.

FLIGHT OFFER:
- Price: ${offer.price.total} ${offer.price.currency}
- Total Duration: ${Math.round(offer.total_duration_minutes / 60)}h ${offer.total_duration_minutes % 60}m
- Number of Stops: ${offer.num_stops}
- Cabin Class: ${offer.class}
- Segments: ${offer.segments.length}
`

  if (offer.segments.length > 0) {
    prompt += `- Departure: ${offer.segments[0].from.city} (${offer.segments[0].from.code}) at ${new Date(offer.segments[0].departure).toLocaleString()}\n`
    prompt += `- Arrival: ${offer.segments[offer.segments.length - 1].to.city} (${
      offer.segments[offer.segments.length - 1].to.code
    }) at ${new Date(offer.segments[offer.segments.length - 1].arrival).toLocaleString()}\n`
  }

  if (offer.layovers.length > 0) {
    prompt += `- Layovers: ${offer.layovers.map((l) => `${l.airport} (${Math.round(l.duration_minutes / 60)}h ${l.duration_minutes % 60}m)`).join(", ")}\n`
  }

  prompt += `- Airlines: ${offer.segments.map((s) => s.airline.name).join(", ")}\n`
  if (offer.notes.length > 0) {
    prompt += `- Notes: ${offer.notes.join(", ")}\n`
  }

  prompt += `\nUSER PREFERENCES:\n`

  if (preferences.budget) {
    prompt += `- Budget: Max ${preferences.budget.max} ${preferences.budget.currency}${preferences.budget.flexible ? " (flexible)" : ""}\n`
  }

  if (preferences.preferred_times?.departure_window) {
    const dw = preferences.preferred_times.departure_window
    if (dw.earliest_hour !== undefined || dw.latest_hour !== undefined) {
      prompt += `- Departure Window: ${dw.earliest_hour || "any"}:00 - ${dw.latest_hour || "any"}:00\n`
    }
    if (dw.preferred_hours?.length) {
      prompt += `- Preferred Departure Hours: ${dw.preferred_hours.join(", ")}:00\n`
    }
  }

  if (preferences.layover_tolerance) {
    const lt = preferences.layover_tolerance
    if (lt.max_layovers !== undefined) {
      prompt += `- Max Layovers: ${lt.max_layovers}\n`
    }
    if (lt.preferred_airports?.length) {
      prompt += `- Preferred Connection Airports: ${lt.preferred_airports.join(", ")}\n`
    }
    if (lt.avoid_airports?.length) {
      prompt += `- Avoid Airports: ${lt.avoid_airports.join(", ")}\n`
    }
  }

  if (preferences.preferred_airlines?.length) {
    prompt += `- Preferred Airlines: ${preferences.preferred_airlines.join(", ")}\n`
  }

  if (preferences.avoid_airlines?.length) {
    prompt += `- Avoid Airlines: ${preferences.avoid_airlines.join(", ")}\n`
  }

  if (preferences.cabin_class) {
    prompt += `- Preferred Cabin Class: ${preferences.cabin_class}\n`
  }

  prompt += `\nEvaluate this flight offer and provide:
1. Overall score (0-100): 100 = perfect match, 0 = terrible match
2. Reasoning: Brief explanation of the score
3. Match details: Scores for budget_match, time_match, layover_match, airline_match, overall_fit (each 0-1)

Consider:
- Price relative to budget
- Departure/arrival times vs preferences
- Layover count and duration vs tolerance
- Airlines vs preferred/avoided lists
- Overall journey quality (duration, stops, cabin class)`

  return prompt
}

/**
 * Fallback scoring when LLM fails
 * Uses simple rule-based scoring
 */
function scoreFlightOfferFallback(offer: FlightOffer, preferences: FlightPreferences): ScoredFlightOffer {
  let score = 50 // Base score
  const details: ScoredFlightOffer["match_details"] = {}

  // Budget match
  if (preferences.budget?.max) {
    const priceRatio = offer.price.total / preferences.budget.max
    if (priceRatio <= 1.0) {
      details.budget_match = 1.0
      score += 20
    } else if (priceRatio <= 1.2 && preferences.budget.flexible) {
      details.budget_match = 0.7
      score += 10
    } else {
      details.budget_match = Math.max(0, 1 - (priceRatio - 1))
      score -= 20
    }
  }

  // Layover match
  if (preferences.layover_tolerance?.max_layovers !== undefined) {
    if (offer.num_stops <= preferences.layover_tolerance.max_layovers) {
      details.layover_match = 1.0
      score += 15
    } else {
      details.layover_match = 0.3
      score -= 15
    }
  }

  // Airline match
  if (preferences.preferred_airlines?.length) {
    const hasPreferred = offer.segments.some((s) =>
      preferences.preferred_airlines!.some((pa) =>
        s.airline.name.toLowerCase().includes(pa.toLowerCase()) || s.airline.code === pa,
      ),
    )
    if (hasPreferred) {
      details.airline_match = 1.0
      score += 10
    }
  }

  if (preferences.avoid_airlines?.length) {
    const hasAvoided = offer.segments.some((s) =>
      preferences.avoid_airlines!.some((aa) =>
        s.airline.name.toLowerCase().includes(aa.toLowerCase()) || s.airline.code === aa,
      ),
    )
    if (hasAvoided) {
      details.airline_match = 0.2
      score -= 20
    }
  }

  // Clamp score to 0-100
  score = Math.max(0, Math.min(100, score))
  details.overall_fit = score / 100

  return {
    ...offer,
    score,
    reasoning: `Rule-based scoring: Budget ${details.budget_match?.toFixed(2) || "N/A"}, Layovers ${details.layover_match?.toFixed(2) || "N/A"}, Airlines ${details.airline_match?.toFixed(2) || "N/A"}`,
    match_details: details,
  }
}

File: _archived/backend_archived/app/services/llm_scorer.py
"""LLM-based flight scoring service."""
import json
import re
from typing import Any
import httpx

from ..core.config import get_settings

settings = get_settings()


async def extract_preferences(holiday: dict, additional_context: str = "") -> dict[str, Any]:
    """Extract user preferences from holiday data using LLM."""
    if not settings.OPENAI_API_KEY:
        return _extract_basic_preferences(holiday)
    
    prompt = f"""Analyze this holiday configuration and extract travel preferences:

Holiday: {json.dumps(holiday, default=str)}
Context: {additional_context}

Return a JSON object with these optional fields:
{{
  "budget": {{"max": number, "currency": "EUR", "flexible": boolean}},
  "preferred_times": {{"departure_window": {{"earliest_hour": 0-23, "latest_hour": 0-23}}}},
  "layover_tolerance": {{"max_layovers": number, "max_layover_minutes": number}},
  "preferred_airlines": ["airline names"],
  "cabin_class": "Economy" | "Business" | "First"
}}

Only include fields you can infer from the data. Return ONLY the JSON object."""

    try:
        async with httpx.AsyncClient(timeout=30.0) as client:
            response = await client.post(
                "https://api.openai.com/v1/chat/completions",
                headers={
                    "Authorization": f"Bearer {settings.OPENAI_API_KEY}",
                    "Content-Type": "application/json",
                },
                json={
                    "model": "gpt-4o-mini",
                    "messages": [{"role": "user", "content": prompt}],
                    "temperature": 0.2,
                },
            )
            response.raise_for_status()
            data = response.json()
            text = data["choices"][0]["message"]["content"]
            
            json_match = re.search(r'\{[\s\S]*\}', text)
            if json_match:
                return json.loads(json_match.group())
                
    except Exception as e:
        print(f"[LLM Scorer] Preference extraction error: {e}")
    
    return _extract_basic_preferences(holiday)


def _extract_basic_preferences(holiday: dict) -> dict[str, Any]:
    """Extract basic preferences without LLM."""
    prefs = {}
    
    if holiday.get("budget"):
        prefs["budget"] = {
            "max": holiday["budget"],
            "currency": "EUR",
            "flexible": True,
        }
    
    if holiday.get("max_layovers") is not None:
        prefs["layover_tolerance"] = {
            "max_layovers": holiday["max_layovers"],
        }
    
    return prefs


async def score_flight_offers(
    offers: list[dict],
    preferences: dict[str, Any],
) -> list[dict]:
    """Score flight offers based on preferences using LLM."""
    if not offers:
        return []
    
    if not settings.OPENAI_API_KEY:
        return _score_basic(offers, preferences)
    
    # Limit offers to avoid token limits
    offers_to_score = offers[:20]
    
    prompt = f"""Score these flight offers based on user preferences.

Preferences: {json.dumps(preferences)}

Offers (simplified):
{json.dumps([{
    "id": o["id"],
    "price": o["price"],
    "num_stops": o["num_stops"],
    "total_duration_minutes": o["total_duration_minutes"],
    "cabin_class": o.get("cabin_class", "Economy"),
} for o in offers_to_score], indent=2)}

For each offer, provide a score (0-100) and brief reasoning.
Return a JSON array:
[
  {{"id": "offer_id", "score": 85, "reasoning": "Good price, direct flight"}}
]"""

    try:
        async with httpx.AsyncClient(timeout=60.0) as client:
            response = await client.post(
                "https://api.openai.com/v1/chat/completions",
                headers={
                    "Authorization": f"Bearer {settings.OPENAI_API_KEY}",
                    "Content-Type": "application/json",
                },
                json={
                    "model": "gpt-4o-mini",
                    "messages": [{"role": "user", "content": prompt}],
                    "temperature": 0.2,
                },
            )
            response.raise_for_status()
            data = response.json()
            text = data["choices"][0]["message"]["content"]
            
            json_match = re.search(r'\[[\s\S]*\]', text)
            if json_match:
                scores = json.loads(json_match.group())
                score_map = {s["id"]: s for s in scores}
                
                scored_offers = []
                for offer in offers_to_score:
                    score_data = score_map.get(offer["id"], {})
                    scored_offers.append({
                        **offer,
                        "score": score_data.get("score", 50),
                        "reasoning": score_data.get("reasoning", ""),
                        "match_details": {},
                    })
                
                # Sort by score descending
                scored_offers.sort(key=lambda x: x["score"], reverse=True)
                return scored_offers
                
    except Exception as e:
        print(f"[LLM Scorer] Scoring error: {e}")
    
    return _score_basic(offers, preferences)


def _score_basic(offers: list[dict], preferences: dict) -> list[dict]:
    """Basic scoring without LLM."""
    budget_max = preferences.get("budget", {}).get("max")
    max_layovers = preferences.get("layover_tolerance", {}).get("max_layovers")
    
    scored = []
    for offer in offers:
        score = 50  # Base score
        reasoning_parts = []
        
        price = offer.get("price", {}).get("total", 0)
        stops = offer.get("num_stops", 0)
        
        # Price scoring
        if budget_max and price:
            if price <= budget_max:
                score += 20
                reasoning_parts.append("Within budget")
            elif price <= budget_max * 1.1:
                score += 10
                reasoning_parts.append("Slightly over budget")
        
        # Layover scoring
        if stops == 0:
            score += 20
            reasoning_parts.append("Direct flight")
        elif max_layovers is not None and stops <= max_layovers:
            score += 10
            reasoning_parts.append(f"{stops} stop(s)")
        
        scored.append({
            **offer,
            "score": min(100, score),
            "reasoning": "; ".join(reasoning_parts) if reasoning_parts else "Standard option",
            "match_details": {},
        })
    
    scored.sort(key=lambda x: x["score"], reverse=True)
    return scored
